### ISSUE TITLE
PixelShuffle for downscaling. (Feature proposal)

### ISSUE THREAD

**Issue by allanzelener** [pytorch/pytorch#1684]
Currently the PixelShuffle module only implements scaling by an upscaling factor > 1. However an equivalent operation that performs downscaling would involve only a small tweak and is used in some models, e.g. YOLOv2. This would also bring feature parity with Tensorflow which has both `depth_to_space` and `space_to_depth` as equivalent operations. 

This would involve the following change to pixel_shuffle:

```
def pixel_shuffle(input, scale_factor):
    batch_size, in_channels, in_height, in_width = input.size()

    out_channels = channels // (scale_factor * scale_factor)
    out_height = in_height * scale_factor
    out_width = in_width * scale_factor

    if scale_factor >= 1:
        input_view = input.contiguous().view(
            batch_size, channels, upscale_factor, upscale_factor,
            in_height, in_width)
        shuffle_out = input_view.permute(0, 1, 4, 2, 5, 3).contiguous()
    else:
        block_size = 1 / scale_factor
        input_view = input.contiguous().view(
            batch_size, channels, out_height, block_size,
            out_width, block_size)
        shuffle_out = input_view.permute(0, 1, 3, 5, 2, 4).contiguous()

    return shuffle_out.view(batch_size, out_channels, out_height, out_width)
```
Also it doesn't look like pixel_shuffle currently has any error checking to see if the scale factor is valid. `out_channels`, `out_height`, and `out_width` should all be integers. 

cc @albanD @mruberry @jbschlosser

**Comment by fmassa** [comment]
That seems like a nice thing to add.
I have one comment though: I'd maybe enforce that the transform is invertible, so doing `pixel_shuffle(pixel_shuffle(x, 2), 1. / 2) == x`. What do you think?

**Comment by allanzelener** [comment]
@fmassa Agreed. I haven't verified yet if my proposed implementation has this property, need to find some time to thoroughly test it.

**Comment by fmassa** [comment]
I don't think it's in pytorch yet. And I'd be ok with having it inside `pixel_shuffle`. @alykhantejani Any objection on that?

**Comment by alykhantejani** [comment]
I think this would be a great feature. We only used it for upscaling, hence not considering this case - but would be nice to support scale factors < 1. I also agree with @fmassa that it should retain the property `pixel_shuffle(pixel_shuffle(x, 2), 1. / 2) == x`.

We should also add this into the lua-torch version too

**Comment by mohammad-py** [comment]
One thing to note, `pixelshuffle` with down scale feature is not as same as *tf.space_to_depth**. Above implementation does shuffling, but `tf.space_to_depth` does not! It is just a reshaping function.

**Comment by zou3519** [comment]
As per @fmassa's comment, we'd be happy to review and accept a PR that implemented this.

**Comment by liaomingg** [comment]
In [10]: x = torch.randn((1, 4, 2, 2))                                                                                                                                                                    

In [11]: x                                                                                                                                                                                                
Out[11]: 
tensor([[[[ 1.1784,  0.2635],
          [-0.9363, -0.6162]],

         [[ 0.2314, -0.4491],
          [-1.1700, -1.3934]],

         [[-0.9268,  0.2151],
          [ 0.9150, -0.1797]],

         [[ 0.0940,  0.9673],
          [-0.8963, -0.3925]]]])

In [12]: y = torch.pixel_shuffle(x, 2)                                                                                                                                                                    

In [13]: y                                                                                                                                                                                                
Out[13]: 
tensor([[[[ 1.1784,  0.2314,  0.2635, -0.4491],
          [-0.9268,  0.0940,  0.2151,  0.9673],
          [-0.9363, -1.1700, -0.6162, -1.3934],
          [ 0.9150, -0.8963, -0.1797, -0.3925]]]])

In [14]: def space_to_depth(x, downscale_factor): 
    ...:     '''''' 
    ...:     batch_size, in_channels, in_height, in_width = x.size() 
    ...:     channels = in_channels * downscale_factor ** 2 
    ...:     out_height = in_height // downscale_factor 
    ...:     out_width = in_width // downscale_factor  
    ...:  
    ...:     x_view = x.contiguous().view( 
    ...:         batch_size, in_channels, downscale_factor, downscale_factor,  
    ...:         out_height, out_width 
    ...:     ) 
    ...:  
    ...:     shuffle_out = x_view.permute(0, 1, 3, 5, 2, 4).contiguous() 
    ...:     return shuffle_out.view(batch_size, channels, out_height, out_width) 
    ...:                                                                                                                                                                                                  

In [15]: z = space_to_depth(y, 2)                                                                                                                                                                         

In [16]: z                                                                                                                                                                                                
Out[16]: 
tensor([[[[ 1.1784,  0.2635],
          [-0.9363, -0.6162]],

         [[ 0.2314, -0.4491],
          [-1.1700, -1.3934]],

         [[-0.9268,  0.2151],
          [ 0.9150, -0.1797]],

         [[ 0.0940,  0.9673],
          [-0.8963, -0.3925]]]])

**Comment by jbschlosser** [comment]
Added in #49334
